### PR TITLE
feature/authorization: Add auth middleware to implement RBAC

### PR DESCRIPTION
**Pull request by ayxsth**
## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] I have created postman test scripts for it

### CODE REVIEW

File: src/middlewares/authenticateToken.js [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import jwt from 'jsonwebtoken';
 import Boom from '@hapi/boom';
+import User from '@/models/user';
```
Comment (sbimochan) [inline src/middlewares/authenticateToken.js]:
```suggestion

import User from '@models/user';
```

File: src/middlewares/authenticateToken.js [hunk 2]
```diff
@@ -19,11 +20,26 @@ export function authenticateToken(req, res, next) {
...
+    }
+
+    const userData = user.filterSensitiveData();
+
+    // Include roles in the response
+    const roles = user.related('roles').toJSON();
+
+    userData.roles = roles;
+
+    req.user = userData;
```
Comment (sbimochan) [inline src/middlewares/authenticateToken.js]:
```suggestion
	  const userData = user.filterSensitiveData();
	
	  // Attach roles to userData
	  req.user = {
	    ...userData,
	    roles: user.related('roles').toJSON(),
	  };
```
Also, filterSensitiveData already serializes to JSON. can we have single toJSON to serialize at once?

File: src/middlewares/authorize.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import Boom from '@hapi/boom';
+
+import { permissions } from '@/constants/permission';
```
Comment (sbimochan) [inline src/middlewares/authorize.js]:
```suggestion
import { permissions } from '@constants/permission';
```

File: src/middlewares/authorize.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    const { user } = req;
+
+    if (!user) {
+      return next(Boom.unauthorized('Unauthorized'));
+    }
+
+    const { roles } = user;
+
+    const userPermissions = [...new Set(...roles.map((role) => permissions[role.name]))];
```
Comment (sbimochan) [inline src/middlewares/authorize.js]: let's not convert set to array. let's keep on using set.
can you try

```javascript
    const userPermissions = roles.flatMap(role => permissions[role.name]);
    const uniquePermissions = new Set(userPermissions);

    if (!uniquePermissions.has(requiredPermission)) {
```

**Review by Abhigyandhakallf — APPROVED** [review]
Looks good

File: src/routes/rentalRoutes.js [hunk 5]
```diff
@@ -2,19 +2,21 @@ import { Router } from 'express';
 
 import * as rentsController from '@controllers/rentsController';
 
+import { authorize } from '@/middlewares/authorize';
```
Comment (sbimochan) [inline src/routes/rentalRoutes.js]:
```suggestion
import { authorize } from '@middlewares/authorize';
```

File: src/routes/roleRoutes.js [hunk 6]
```diff
@@ -2,28 +2,30 @@ import { Router } from 'express';
 
 import * as roleController from '@controllers/rolesController';
 
+import { authorize } from '@/middlewares/authorize';
```
Comment (sbimochan) [inline src/routes/roleRoutes.js]:
```suggestion
import { authorize } from '@middlewares/authorize';
```

File: src/routes/roleRoutes.js [hunk 7]
```diff
@@ -2,28 +2,30 @@ import { Router } from 'express';
...
  * POST /api/roles.
  */
-router.post('/', roleValidator, roleController.create);
+router.post('/', authorize('role.create'), roleValidator, roleController.create);
 
 /**
  * PUT /api/roles/:id.
  */
-router.put('/:id', findRole, roleValidator, roleController.update);
+router.put('/:id', authorize('role.update'), findRole, roleValidator, roleController.update);
```
Comment (sbimochan) [inline src/routes/roleRoutes.js]: I'm feeling like too many middleewares. how about putting in service layer?

File: src/routes/userRoleRoutes.js [hunk 8]
```diff
@@ -2,6 +2,8 @@ import { Router } from 'express';
 
 import * as userRoleController from '@controllers/userRoleController';
 
+import { authorize } from '@/middlewares/authorize';
```
Comment (sbimochan) [inline src/routes/userRoleRoutes.js]:
```suggestion
import { authorize } from '@middlewares/authorize';
```

File: src/routes/userRoutes.js [hunk 9]
```diff
@@ -4,34 +4,36 @@ import userRoleRoutes from './userRoleRoutes';
 
 import * as usersController from '@controllers/usersController';
 
+import { authorize } from '@/middlewares/authorize';
```
Comment (sbimochan) [inline src/routes/userRoutes.js]:
```suggestion
import { authorize } from '@middlewares/authorize';
```

**Comment by sbimochan** [comment]
@ayxsth please update the integration tests too

**Review by sbimochan — APPROVED** [review]

**Pull request merged**